File: examples/nodejs/basic_to_medium.mjs
/**
 * Basic → Medium (Node.js / @veristamp/nqql)
 *
 * Offline walkthrough: parse, isValid, explain, compileQuery,
 * injectFilter, injectShardKey, hybrid shorthand.
 */
import nqql from '../../crates/nqql/index.js';

const {
  parse,
  isValid,
  explain,
  compileQuery,
  injectFilter,
  injectShardKey,
  version,
} = nqql;

console.log(`nqql ${version ?? nqql.__version__ ?? '?'}\n`);

const q = "QUERY TEXT 'machine learning' FROM papers USING dense LIMIT 5";

// 1. Validate & parse
console.log('1. isValid:', isValid(q));
const [stmt] = parse(q);
console.log(
  '   Stmt methods:',
  Object.getOwnPropertyNames(Object.getPrototypeOf(stmt))
    .filter((n) => n !== 'constructor')
    .join(', '),
);
console.log();

// 2. Explain
console.log('2. explain()');
console.log(explain(q));
console.log();

// 3. Compile to REST route
const route = compileQuery(q);
console.log('3. compileQuery()');
console.log(`   ${route.stmt_type}  ${route.method} ${route.path}`);
console.log();

// 4. injectFilter (free function → AST object)
const securedAst = injectFilter(q, 'org_id', '=', 'acme-corp');
console.log('4. injectFilter(org_id = acme-corp)');
console.log('   keys:', Object.keys(securedAst));
console.log();

// 5. Stmt.injectFilter + injectShardKey (in-place)
const [s] = parse(q);
s.injectFilter('tenant_id', '=', 'acme-corp');
s.injectShardKey('acme-corp');
console.log('5. Stmt.injectFilter + injectShardKey');
console.log('   shardKey =', s.shardKey);
console.log();

// 6. Hybrid shorthand
const hybrid = `
  QUERY TEXT 'transformers attention'
  FROM papers
  USING HYBRID DENSE dense SPARSE sparse FUSION RRF
  LIMIT 5
`;
console.log('6. hybrid shorthand isValid:', isValid(hybrid));
console.log(explain(hybrid));

File: examples/python/basic_to_medium.py
#!/usr/bin/env python3
"""
Basic → Medium (Python / pyqql)

Offline walkthrough:
  1. parse / is_valid
  2. explain
  3. compile_query
  4. inject_filter (logical isolation)
  5. SHARD in QQL + Stmt.shard_key (physical routing)
  6. hybrid shorthand
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path

_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, os.environ.get("QQL_LIB", str(_ROOT / "crates" / "pyqql")))

import pyqql  # noqa: E402

print(f"pyqql {getattr(pyqql, '__version__', '?')}\n")

q = "QUERY TEXT 'cardiology treatment protocols' FROM medical_records USING dense LIMIT 5"
print("1. is_valid:", pyqql.is_valid(q))
print("   parsed Stmt OK\n")

print("2. explain()")
print(pyqql.explain(q).get("plan", ""))
print()

route = pyqql.compile_query(q)
print("3. compile_query()")
print(f"   stmt_type={route.get('stmt_type')}  {route.get('method')} {route.get('path')}\n")

# Fast JSON-only parse — same AST as parse(), no Python Stmt objects.
print("3b. parse_json()")
print(f"   {pyqql.parse_json(q)[:120]}…\n")

secured = pyqql.inject_filter(q, "tenant_id", "=", "hospital-east")
print("4. inject_filter(tenant_id = hospital-east)")
print("   filter:", json.dumps(secured.to_dict().get("Query", {}).get("filter"), indent=2)[:300])
print()

# Preferred: write SHARD in QQL (security + routing in the language)
sharded_q = """
    QUERY TEXT 'cardiology treatment protocols'
    FROM medical_records
    USING dense
    SHARD 'hospital-east'
    LIMIT 5
"""
stmt = pyqql.parse(sharded_q)[0]
print("5a. SHARD in QQL → shard_key =", repr(stmt.shard_key))
# Host resolve-after-parse (optional): property only, no inject_shard_key
plain = pyqql.parse(q)[0]
plain.shard_key = "hospital-east"
print("5b. Stmt.shard_key = … →", repr(plain.shard_key))
print()

hybrid = """
    QUERY TEXT 'emergency neurological assessment'
    FROM medical_records
    USING HYBRID DENSE dense SPARSE sparse FUSION RRF
    WHERE priority = 'high'
    SHARD 'hospital-east'
    LIMIT 5
"""
print("6. hybrid + SHARD — is_valid:", pyqql.is_valid(hybrid))
print(pyqql.explain(hybrid).get("plan", ""))
print("\nDone.")
